### PR TITLE
Use of Max-Sample algorithm for gain-switch cases in the barrel

### PR DESCRIPTION
**Pull request by emanueledimarco**
This is a fix for the ECAL local reconstruction that, for EB only, where the problem is visible,
changes the algorithm from multifit to the simple max-sample. 
This solution for EB grants the consistency of the (legacy) re-reco data with the Summer16 MC samples already existing [1]. For EE it would introduce a small extra smearing in the data/MC, so it is switched off for the re-reco (since the problem affects only EB). 
This should then be customised to be used for 2015-2016 data only. 

Depending to what studies for 90X with improved multifit show, related to PR #17205, we could either use them, if positive, or fall back on this max-sample solution, for both EB and EE. 
Should this be implemented for 90X as well on top of PR #17205 ?

@bendavid @amassiro @slava77 

[1] https://indico.cern.ch/event/590956/contributions/2391934/attachments/1384258/2105710/multifit-gainswitch.pdf

### CODE REVIEW

**Comment by slava77** [comment]
On 1/24/17 4:15 AM, Emanuele Di Marco wrote:
> Should this be implemented for 90X as well on top of PR #17205
> <https://github.com/cms-sw/cmssw/pull/17205> ?

the nominal policy is to have the features in 90X first

File: RecoLocalCalo/EcalRecProducers/plugins/EcalUncalibRecHitWorkerMultiFit.cc [hunk 1]
```diff
@@ -573,6 +575,7 @@ EcalUncalibRecHitWorkerMultiFit::getAlgoDescription() {
 	      edm::ParameterDescription<bool>("ampErrorCalculation", true, true) and
 	      edm::ParameterDescription<bool>("useLumiInfoRunHeader", true, true) and
 	      edm::ParameterDescription<int>("bunchSpacing", 0, true) and
+              edm::ParameterDescription<bool>("gainSwitchEBMaxSample", true, true) and
```
Comment (slava77) [inline RecoLocalCalo/EcalRecProducers/plugins/EcalUncalibRecHitWorkerMultiFit.cc]: I think that the default here should be "false"

Comment (emanueledimarco) [inline RecoLocalCalo/EcalRecProducers/plugins/EcalUncalibRecHitWorkerMultiFit.cc]: OK, will do

**Comment by emanueledimarco** [comment]
For 90X, I will add it as configurable on top of PR #17205, since it is already merged

**Comment by slava77** [comment]
@cmsbuild please test

@emanueledimarco 
what is needed to enable the feature?
Is it just ``` process.ecalMultiFitUncalibRecHit.gainSwitchEBMaxSample = False ``` ?

**Comment by emanueledimarco** [comment]
@slava77 
what is needed to enable the feature?
Is it just process.ecalMultiFitUncalibRecHit.gainSwitchEBMaxSample = False ?

The opposite: process.ecalMultiFitUncalibRecHit.gainSwitchEBMaxSample = True
only for 2016 data for the time being

**Comment by slava77** [comment]
On 1/24/17 6:42 AM, Emanuele Di Marco wrote:
> @slava77 <https://github.com/slava77>
> what is needed to enable the feature?
> Is it just process.ecalMultiFitUncalibRecHit.gainSwitchEBMaxSample = False ?
> 
> The opposite: process.ecalMultiFitUncalibRecHit.gainSwitchEBMaxSample = True
> only for 2016 data for the time being
> 

Thanks.
It is "True" indeed, I copy-pasted too much without thinking.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by emanueledimarco** [comment]
Forgot to add the validation with 2016 data (from Shervin Nourbakhsh and Giuseppe Fasanella)
@shervin86 

[multifit-gainswitch (1).pdf](https://github.com/cms-sw/cmssw/files/726998/multifit-gainswitch.1.pdf)

**Comment by slava77** [comment]
@cmsbuild please test
blunt attempt to rerun
both problems are due to file read problems on xrootd side
if it repeats, chances are that the inputs disappeared and need to be pulled back on accessible disk/site

File: RecoLocalCalo/EcalRecProducers/python/ecalMultiFitUncalibRecHit_cfi.py [hunk 2]
```diff
@@ -14,7 +14,8 @@
       activeBXs = cms.vint32(-5,-4,-3,-2,-1,0,1,2,3,4),
       ampErrorCalculation = cms.bool(True),
       useLumiInfoRunHeader = cms.bool(True),
-  
+
+      gainSwitchEBMaxSample = cms.bool(False),
```
Comment (slava77) [inline RecoLocalCalo/EcalRecProducers/python/ecalMultiFitUncalibRecHit_cfi.py]: what's a good reason for using different names and a different set of switches for the same thing in 80X and 90X? ```gainSwitchEBMaxSample``` in this PR (80X) vs ```gainSwitchUseMaxSampleEB``` and ```gainSwitchUseMaxSampleEE``` in #17279 (90X)

Comment (emanueledimarco) [inline RecoLocalCalo/EcalRecProducers/python/ecalMultiFitUncalibRecHit_cfi.py]: in 90X I have used a notation similar to the other customization methods. 
For consistency I will change the notation of 80X branch to be the same as in 90X.

**Comment by emanueledimarco** [comment]
with this naming convention, for 80X legacy re-reco, the customisation line is:

process.ecalMultiFitUncalibRecHit.algoPSet.gainSwitchUseMaxSampleEB = cms.bool(True)

**Comment by slava77** [comment]
@cmsbuild please test

@emanueledimarco thanks
it took me a few attempts to run in 90X this weekend having a name from 80X from my inbox until I noticed that the names don't match.

**Comment by slava77** [comment]
backport of #17279 (functional; not verbatim due to some changes in context)

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/17279	455267f
- implementation is consistent with 90X version #17279; the defaults are not changed, the new behavior should be enabled with a config switch
- jenkins tests pass and comparisons with baseline show no difference
- relying on conclusions made in #17279 to sign off on the added functionality

**Comment by slava77** [comment]
@emanueledimarco 
if we want the gainswitch fix to be enabled in 80X rereco, it will have to be tested or become a default in 90X, even if it is soon to be replaced by more elaborate treatment in the multifit.

Since multifit is applicable to all eras, this switch can be added to 
Configuration/StandardSequences/python/Reconstruction_Data_cff.py to become a default.
Please make a corresponding 90X PR

**Comment by rafaellopesdesa** [comment]
@emanueledimarco @slava77 What was the end of this PR in 80X? If I run the ECAL reco in 80X will it use the max-amp for hits with gain 6 and 1 (I mean, was there another PR that I missed and it was merged?)

**Comment by slava77** [comment]
On 2/16/17 6:42 AM, Rafael Lopes de Sa wrote:
> @emanueledimarco <https://github.com/emanueledimarco> @slava77
> <https://github.com/slava77> What was the end of this PR in 80X? If I
> run the ECAL reco in 80X will it use the max-amp for hits with gain 6
> and 1 (I mean, was there another PR that I missed and it was merged?)

80X RECO is still unchanged, until shortly before we are ready to build
the rereco release.

**Comment by rafaellopesdesa** [comment]
@slava77 Thanks for the info.

**Comment by emanueledimarco** [comment]
I have included the commits of @bendavid of #17876 to make a unique one for 80X
@amassiro @argiro

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/17259	c8b69f0
- updated "max-sample" choice for the gain-switched hit reco; the default behavior is still unchanged. Changes are in sync with https://github.com/cms-sw/cmssw/pull/17877 9fbc630
- jenkins tests pass
- sign off relies on checks done in https://github.com/cms-sw/cmssw/pull/17877 9fbc630

**Comment by emanueledimarco** [comment]
This is still pending and should be:

- included in 8.0.X 
- turned ON for the 2016 legacy re-reco only for the ECAL barrel, i.e.: process.ecalMultiFitUncalibRecHit.algoPSet.gainSwitchEBMaxSample = cms.bool(True)

@paramatti @fcouderc also follow this.

**Comment by amassiro** [comment]
As discussed today at the AlCa meeting (https://indico.cern.ch/event/626460/), can this be switched on for the legacy re-reco cmssw release?
Cheers, ECAL

**Comment by slava77** [comment]
On 4/5/17 4:37 PM, Andrea Massironi wrote:
> As discussed today at the AlCa meeting
> (https://indico.cern.ch/event/626460/), can this be switched on for the
> legacy re-reco cmssw release?
> Cheers, ECAL
> 

we can still preserve data-MC consistency by keeping the default off
and will enable this feature in the workflow configuration using

--customise_command
"process.ecalMultiFitUncalibRecHit.algoPSet.gainSwitchEBMaxSample =
cms.bool(True)\n"

**Comment by amassiro** [comment]
Perfect. Thanks a lot for fast reply.

**Comment by emanueledimarco** [comment]
@slava77 ok. But this should be merged in 8.0.X still, right?

**Comment by slava77** [comment]
On 4/5/17 4:57 PM, Emanuele Di Marco wrote:
> @slava77 <https://github.com/slava77> ok. But this should be merged in
> 8.0.X still, right?
> 

yes

**Comment by davidlange6** [comment]
hi @emanueledimarco - please submit a PR turning this feature on by default. Thanks

**Comment by emanueledimarco** [comment]
@davidlange6 I have understood from @slava77 in https://github.com/cms-sw/cmssw/pull/17259#issuecomment-291883594 that we want to leave the default off and turn on this feature by a customise command on the legacy re-reco in 8.0.X. 
Is this correct?

**Comment by davidlange6** [comment]
would prefer to make the default correspond to the "legacy" (or at least the first legacy)

**Comment by emanueledimarco** [comment]
fair. Doing it now.

**Pull request merged**